Foundation Encoders

itaiferber opened the pull request:
<!-- What's in this pull request? -->
This implements the `JSONEncoder`, `JSONDecoder`, `PropertyListEncoder`, and `PropertyListDecoder` types proposed by [SE-0167](https://github.com/apple/swift-evolution/blob/master/proposals/0167-swift-encoders.md).

This pulls all Foundation-specific work out of #8124 for separate integration.

**This PR depends on #9004 and should NOT be merged until #9004 is complete and merged.**
rdar://problem/30471969
rdar://problem/30472028

Commit:
Add JSON and property list encoders and decoders

* Integrate {JSON,PropertyList}{Encoder,Decoder} types to facilitate
  encoding types in JSON and property list formats
* Adds Foundation-specific extensions to allow errors exposed from the
  stdlib to bridge to NSErrors

DougGregor commented:
@swift-CI please test

itaiferber commented:
@DougGregor This isn't quite ready to be merged yet — I've got unit tests incoming either later tonight or early tomorrow. But when they make it I'll test and ping you if ready to merge

DougGregor commented:
Ok!

Commit:
Add Codable conformance to Array

Commit:
Ensure all Dates can always round-trip by default

Commit:
Remove unnecessary CodingKey implementation

Commit:
Use encode(contentsOf:) instead of a loop

Commit:
Add preliminary {JSON,Plist}Encoder tests

itaiferber commented:
@swift-ci Please test

DougGregor commented:
@swift-ci please clean test

tkremenek commented:
@swift-ci clean test macOS

itaiferber commented:
@tkremenek I think the linker errors are legitimate, no? The box types that I added were marked as `fileprivate` where I think they should actually be `internal` (I looked at the box types for `Sequence`, `Iterator`, etc. and they are all `@_fixed_layout @_versioned internal`). I'm about to push that change and test

tkremenek commented:
@itaiferber great point; I wasn't certain if those failures were genuine or not.  I'll cancel the job I just initiated.

Commit:
Fix encoding container linker errors

The box types were previously fileprivate because they lived in
the Foundation overlay. As part of the Swift stdlib, though, they need
to be internal so they can be linked against.

itaiferber commented:
@swift-ci Please test

itaiferber commented:
🎉 @DougGregor @tkremenek When either of you gets a chance, do you mind merging this in please? Thanks!

itaiferber commented:
@tkremenek Thanks!

DougGregor commented:
Congrats to @itaiferber for getting this landed

itaiferber commented:
Thanks, @DougGregor! Looking forward to feedback when people start using it. :)

norio-nomura commented:
@itaiferber I have another question.
What should `codingPath` contain on processing `super` and `nestedContainer`?
Current implementation seems not to contain both of `super` and `key` for `nestedContainer`.

itaiferber commented:
@norio-nomura This is another bug: `codingPath` on nested containers and referencing encoders should include the `codingPath`s of their parents. I have a branch with a fix (https://github.com/itaiferber/swift/tree/foundation-codingpath-fixes) but haven't gotten the chance to put it up as a PR yet. Will do in the next few days.

norio-nomura commented:
@itaiferber Thank you for clarifying. 🙏 
FYI, I'm trying to implement `YAMLEncoder` and `YAMLDecoder` for SE-0166 at https://github.com/jpsim/Yams/pull/43.
I'll check your branches to fix `codingPath` on that.

Dimillian commented:
Is Codable support for NSKeyedArchiver part of another PR ? It's in the SE but can't seems to find it in code for now.

itaiferber commented:
@Dimillian That integration hasn't been completed yet. This will be coming in in a future PR.

bubski commented:
@itaiferber `JSONDecoder` & `JSONEncoder` are not yet available on Linux, or am I missing something?

CodaFi commented:
@bubski How recent a snapshot of Swift are you using?

bubski commented:
@CodaFi I tried both
`swift-4.0-DEVELOPMENT-SNAPSHOT-2017-06-06-a-ubuntu16.04`
and
`swift-4.0-DEVELOPMENT-SNAPSHOT-2017-06-07-a-ubuntu16.04`

CodaFi commented:
It seems that because this has landed in the Darwin SDK that it will need to explicitly be put into Foundation for it to be usable on Linux.  The proposal does not make mention of platform lock-in, so I think this will need to be mirrored over.

itaiferber commented:
@bubski As @CodaFi mentions, this will need to be introduced into [swift-corelibs-foundation](https://github.com/apple/swift-corelibs-foundation) in order to be available on Linux. The code as-is should be easy to transfer over to there without requiring many changes (if at all), but we haven't gotten the chance to (and it seems that no one from the community has either).

CodaFi commented:
I've filed [SR-5195](https://bugs.swift.org/browse/SR-5195) about this.

bubski commented:
@itaiferber I started hacking around the idea
http://github.com/bubski/JSONEncoderLinux
and managed to get it running on Linux.

But I don't have a good `swift-corelibs-foundation` dev environment to do proper implementation (that's why I copy-pasted some internals to expose them).

I'll try to clean it up and make a PR to `swift-corelibs-foundation` so it's easier for someone to verify.

itaiferber commented:
@CodaFi Thanks for that!

itaiferber commented:
@bubski Cool, thanks! The one annoyance will be keeping these things in sync. There are still some outstanding changes that will be landing in `JSONEncoder`/`Decoder` that will need to be ported over as well.

bubski commented:
@itaiferber Yea, I figured. I'm open to any suggestions.
It's also tricky because of discrepancies in the way `Foundation` types casting works on Linux (or rather the fact that it doesn't) together with `JSONSerialization` producing different types on Linux vs macOS.

itaiferber commented:
@bubski Yep, there are a few places where I've noted (in the Darwin implementation) places which may have this discrepancy. I think overall, it might be worth updating `JSONSerialization` to follow suit, but that will require some further review which we simply haven't had time for.

As for keeping things in sync, there's not much at the moment, unfortunately. We struggle with this ourselves. I think we need a general process/solution for improving this.

Dimillian commented:
@itaiferber Is having a default value + support optional Codable property something planet? 

Let me explain, you define a var as such in a Codable object: 

`var progressMax: Int = 0`

But then if you decode your object with a JSON that does not contain progressMax it'll throw a missing key error. Which make sense, but it also make sense to have a default, hardcoded value + a JSON defined one if available.

itaiferber commented:
@Dimillian It's possible to support, but I don't know if that's something we'd want to do out of the box. It's impossible for us to tell whether the default value is assigned for the benefit of other initializers, or whether you meant for it to be a default value if the JSON couldn't be read. I don't know if we can read into the intent there — did you want us to assign a default, or were you expecting the error? No real way for us to know.

Dimillian commented:
@itaiferber In this specific case, the JSON can be different and still represent the same object once decoded. 
If you have a model defined as such

```
class Book: Codable {

var title: String
var author: String
var progress: Int = 0

}
```

The JSON will always have the title and author value defined, but the progress value will be in only in some cases (let's say it's an API and the server control it). 
So it make sense in our case to work with non optional value, because we decided that when progress is not present, it should be equal to 0. 

Is there a way in current implementation to represent that? 

But I understand that you can't really define our intent at a lower level, and in make sense for other use case to throw an error.

itaiferber commented:
@Dimillian You've got a few options here (of course, not all perfect):

1. Make `progress` an `Int?` — this will set it to `nil` by default instead of throwing an error, and you can treat `nil` as "no progress". Of course, this is a subpar experience, but it's the easiest solution.
2. Make `progress` a custom `Progress` type which can initialize from `nil` and has an `Int` raw value. Also not optimal, but it will work.
3. Since you're looking for custom behavior, you can always give a custom `init(from:)` which will call `decodeIfPresent(_:forKey:)` instead of `decode(_:forKey:)` and give a default value there of 0.

None of these solutions are great, but they are at least explicit. I'll think on this some more; perhaps there is a cleaner way to improve the experience of doing this.

Dimillian commented:
@itaiferber Actually, your 1. is what we did. It's a bit subpar because we have to unwrap it each time yes. But that's acceptable. In our case we have a lot of var like that now, as we converted NSCoding/Reflection code to Codable. And we could write custom decoder and encoder but in that case we lose a bit of the benefit of Codable.
I think this should be an option in the decoding strategy of the JSONDecoder. Like an optional decoding strategy, you set it to what you want. If that feasible. Maybe I'll take a look at it.

mxngls commented:
> @Dimillian That integration hasn't been completed yet. This will be coming in in a future PR.

I am currently investigating the Foundation framework's source and cannot find the mentioned support for the `Codable` protocol there. It seems that actual support for archiving and unarchiving using `Codable` landed in [PR 11081](https://github.com/swiftlang/swift/pull/11081) but these changes do not seem to have been properly interpreted for Foundation.

Am I missing something here?

itaiferber commented:
@mxngls Good question. The original support for this did land as part of #11081, and this was back at a time when mixed Obj-C+Swift system frameworks couldn't really exist, so the "Swift part" of Foundation lived as an "overlay" in the Swift project itself that got applied when Foundation was imported in Swift. (_Very_ rough approximation of the specifics here.)

When it was possible for overlays to be migrated down to be part of the respective Obj-C codebases themselves, they were removed from the open-source Swift project and migrated down to their closed-source counterparts. The reason you can't find the source these days is

1. `NSKeyedArchiver` and `NSKeyedUnarchiver` are not (yet) available as part of [swift-foundation](https://github.com/swiftlang/swift-foundation), and still live in the closed-source Obj-C framework
2. I don't believe these extensions ever made their way to [swift-corelibs-foundation](https://github.com/swiftlang/swift-corelibs-foundation) for use on non-Darwin platforms

Though, I am pretty far removed these days, and could very well be wrong. @parkera or @jmschonfeld would know best.

mxngls commented:
@itaiferber Thanks a lot for the swift reply.

Is there any documentation available on how NSKeyedArchiver is implemented, particularly how it compares to the Obj-C NSCoding-based architecture? I am currently in the midst of creating a library in another compiled language that provides archiving capabilities identical to those offered by NSKeyedArchiver and NSKeyedUnarchiver. I aim for feature parity with the implementation found in the open-source Foundation framework and try to keep the library internals as close to the open-source implementation as possible. 

My question is largely wether we need to implement the secure encoding mechanism, as it seems to be unique to the constraints imposed by the runtime environment that dictates how class elements are instantiated, and thus doesn't transfer cleanly to compiled languages like Swift. I have no prior knowledge to neither working with Swift nor Obj-C so please correct me if my understanding here is wrong.

> NSKeyedArchiver and NSKeyedUnarchiver are not (yet) available as part of swift-foundation, and still live in the closed-source Obj-C framework

Could you specify what you mean here? Both parts of the archiver are actually already in [swift-corelibs-foundation](https://github.com/swiftlang/swift-corelibs-foundation/blob/main/Sources/Foundation/NSKeyedArchiver.swift). What is missing is the Codable-based implementation.

itaiferber commented:
@mxngls I'm more than happy to answer your questions, but so as to not have a long conversation in the comments on a 9-year-old PR, do you mind reaching out over email? (Email in my GH bio) It seems there may be some misconceptions worth clearing up.

jmschonfeld commented:
Yeah I believe @itaiferber is correct that these extensions to `NSKeyed(Un)Archiver` were not implemented in swift-corelibs-foundation for `Foundation` on non-Darwin and that's the reason they aren't available. @mxngls Feel free to file a GitHub issue on https://github.com/swiftlang/swift-corelibs-foundation about the lack of parity there and we can investigate whether it's feasible to bring that API to swift-corelibs-foundation where those coders live
